src/bootstrap/sanity.rs

```diff
@@ -92,10 +92,31 @@ pub fn check(build: &mut Build) {
                     .unwrap_or(true)
             })
             .any(|build_llvm_ourselves| build_llvm_ourselves);
-    let need_cmake = building_llvm || build.config.any_sanitizers_enabled();
-    if need_cmake {
-        cmd_finder.must_have("cmake");
-    }
+    let need_default_cmake = if building_llvm || build.config.any_sanitizers_enabled() {
+        let cmake_env_var = env::var_os("CMAKE");
+        if let Some(explicit_name) = build.config.cmake.take() {
+            if let Some(cmake_from_env) = cmake_env_var {
+                if explicit_name != cmake_from_env {
+                    eprintln!(
+                        "env var CMAKE = {cmake_from_env:?} != {explicit_name:?} from config.toml"
+                    )
+                }
+            }
+            build.config.cmake = cmd_finder.must_have(explicit_name).into();
+            None
+        } else {
+            // _very_ simplified version of getenv_target_os("CMAKE") from
+            // https://docs.rs/cmake/0.1.48/src/cmake/lib.rs.html#515
+            if let Some(cmake_from_env) = cmake_env_var {
+                cmd_finder.must_have(cmake_from_env)
+            } else {
+                cmd_finder.must_have("cmake")
+            }
+            .into()
+        }
+    } else {
+        None
+    };
```

Comment:
Most of this logic probably belongs in config.rs, not here. We should always have build.config.cmake as the source of truth for the cmake path, and support setting it only via config.toml. (If it is set in the environment and not equal to the config.toml value, it should be a panic! at config parse time).

Here, we should just have the `if need_cmake { cmd_finder.must_have(&build.config.cmake); }`, with that defaulting to just "cmake" if nothing is set in the config.toml.

Reply:
I have a couple of comments, let me start with minor ones (since major ones are going to take more time to write)

1. > (If it is set in the environment and not equal to the config.toml value, it should be a panic! at config parse time)
    
    Last time, you wrote
    https://github.com/rust-lang/rust/pull/85728#discussion_r709390323
    > If it is set and different, I think we should probably prefer config.toml still but issue a warning (just eprintln! is fine).

    Warn or panic? Let me know of your decision.

2. > We should always have build.config.cmake as the source of truth for the cmake path, and support setting it only via config.toml.

    Are you suggesting to set `CMAKE` env var unconditionally based on what's in `config.toml` (or its default value defined in `config.rs` > `define_config!`)? Is that your opinion? You don't mind breaking a hypothetical existing build env which relies on a custom `CMAKE` env var (which is passed through to `cmake` crate as of today)? Let's say someone has `CMAKE=mycmake` as of today, expecting it to be passed through to `cmake` crate, and your suggestion is that they should set `cmake = "mycmake"` instead, and if they don't, their build should fail with a panic saying sth like "cmake from config.toml wants to use `cmake` but you have a contradictory env var CMAKE=mycmake`"?

    Caveat: from what I see, `cmake` crate accepts a *very* wide variety of env vars,
    https://docs.rs/cmake/0.1.48/src/cmake/lib.rs.html#857-861
    like `CMAKE_x86_64-unknown-linux-gnu`, and I don't think it's realistic for `bootstrap` to cover all of them.

I'll find time to add more

Reply:
@Mark-Simulacrum

Reply:
> Warn or panic? Let me know of your decision.

I think panic! is probably best.

> [...] cmake env

Looking into our use of cmake a little more, it looks like we only ever actually invoke it through the cmake crate, so if users are supplying configuration to us via env variables, the cmake crate would already be picking that up appropriately. In sanity checking, we do use "cmake" without doing the same env-variable lookup that the cmake crate does. (Target-specific *cmake* binaries do not feel particularly likely to be actually necessary, as it's really a host tool).

I'm not too worried about env-variables overriding bootstrap's configuration in the cmake crate - that seems unlikely to be an issue in practice, given how bootstrap itself isn't using cmake directly (except for sanity checks). Given that CMAKE isn't a "standard" env variable (or at least I've seen no evidence of that), I'm not really worried about supporting setting it as an alternative to editing config.toml, either.

So I think the value of config.cmake probably comes down to:

* config.toml, if set
* cmake, if not

After we load that, we want to put it in our own environment (set_var("CMAKE")) so that the cmake crate picks it up, at least as things stand today. Just before doing that I think it makes sense to check that it's not set to a different value (mostly just as a sanity check, but it doesn't make sense for this to be in sanity.rs).

Pretty much all of that logic should go into config.rs.